packages/apollo-engine-reporting/src/treeBuilder.ts

```diff
@@ -0,0 +1,250 @@
+import {
+  GraphQLResolveInfo,
+  GraphQLError,
+  ResponsePath,
+  responsePathAsArray,
+} from 'graphql';
+import { Trace, google } from 'apollo-engine-reporting-protobuf';
+
+export class EngineReportingTreeBuilder {
+  private rootNode = new Trace.Node();
+  public trace = new Trace({ root: this.rootNode });
+  public startHrTime?: [number, number];
+  private stopped = false;
+  private nodes = new Map<string, Trace.Node>([
+    [rootResponsePath, this.rootNode],
+  ]);
+  private rewriteError?: (err: GraphQLError) => GraphQLError | null;
+
+  public constructor(options: {
+    rewriteError?: (err: GraphQLError) => GraphQLError | null;
+  }) {
+    this.rewriteError = options.rewriteError;
+  }
+
+  public startTiming() {
+    if (this.startHrTime) {
+      throw Error('startTiming called twice!');
+    }
+    if (this.stopped) {
+      throw Error('startTiming called after stopTiming!');
+    }
+    this.trace.startTime = dateToProtoTimestamp(new Date());
+    this.startHrTime = process.hrtime();
+  }
+
+  public stopTiming() {
+    if (!this.startHrTime) {
+      throw Error('stopTiming called before startTiming!');
+    }
+    if (this.stopped) {
+      throw Error('stopTiming called twice!');
+    }
+
+    this.trace.durationNs = durationHrTimeToNanos(
+      process.hrtime(this.startHrTime),
+    );
+    this.trace.endTime = dateToProtoTimestamp(new Date());
+    this.stopped = true;
+  }
+
+  public willResolveField(info: GraphQLResolveInfo): () => void {
+    if (!this.startHrTime) {
+      throw Error('willResolveField called before startTiming!');
+    }
+    if (this.stopped) {
+      throw Error('willResolveField called after stopTiming!');
+    }
+
+    const path = info.path;
+    const node = this.newNode(path);
+    node.type = info.returnType.toString();
+    node.parentType = info.parentType.toString();
+    node.startTime = durationHrTimeToNanos(process.hrtime(this.startHrTime));
+    if (typeof path.key === 'string' && path.key !== info.fieldName) {
+      // This field was aliased; send the original field name too (for FieldStats).
+      node.originalFieldName = info.fieldName;
+    }
+
+    return () => {
+      node.endTime = durationHrTimeToNanos(process.hrtime(this.startHrTime));
+    };
+  }
+
+  public didEncounterErrors(errors: GraphQLError[]) {
+    errors.forEach(err => {
+      // This is an error from a federated service. We will already be reporting
```

Comment:
I suppose if we supported nested gateway levels (which won't work now for reporting anyway) then maybe this is wrong. Maybe it should be moved to EngineReportingAgent instead of the tree builder, before the call to the tree builder.